[jaws.bar] 예약 대기 관리 step 1, 2

park0503 opened the pull request:
## 구현 기능
- [step1 요구 사항](https://edu.nextstep.camp/s/GiKTqpMP/ls/ZuVOiQlB) 에 맞춰 auth 패키지를 분리했습니다.
  - `UserDetailsDao` 인터페이스를 구현하고 nextstep 패키지 내의 `memberDao` 클래스가 이를 구현하도록 해 줌으로써 의존성을 역전시켰습니다.
  - auth 패키지 내의 Bean들을 `AuthConfiguration` 클래스에서 등록해 주었습니다.
- [step2 요구 사항](https://edu.nextstep.camp/s/GiKTqpMP/ls/MlRKpeWQ) 에 맞춰 예약 대기 신청 기능을 구현하였습니다.
  - ReservationWaiting 이라는 도메인을 추가했습니다.
  - 추가한 기능에 대한 테스트를 작성했습니다.
-  그 외 스켈레톤 코드에 관해 기본적인 리팩토링을 진행했습니다.
   - 지난 Spring 과제에서 계속 사용해 왔던 GlobalExceptionHandler와 ErrorCode, ErrorResponse를 도입했습니다.
   - `javax.validation` 라이브러리를 이용해 기본적인 입력값 validation을 진행했습니다.
   - 로직에서 부족한 검증 부분 (스케줄 삭제 전 예약 확인 등)을 추가했습니다.
   - 변경된 부분에 따라 테스트를 전면적으로 수정했습니다.

## 기타
- 벌써 스프링 마지막 과제네요! 지난주 지지난주 휴일이 많았던 터라 조금 느슨해져 있었는데 다시 마음 가다듬고 달려보겠습니다. 감사합니다!

park0503 committed:
feat: auth 패키지 분리 및 의존성 최적화

park0503 committed:
feat: 예약 대기 기능 구현

park0503 committed:
test: 예약 대기 API 테스트 작성

park0503 committed:
feat: GlobalExceptionHanlder 도입

- ErrorCode 및 ErrorResponse 도입
- Auth 패키지 내 AuthorizationException 도입
- GlobalExceptionHanlder 내 Auth 패키지 Exception 처리 추가

park0503 committed:
refactor: Member Role Enum 적용

park0503 committed:
feat: 요청값 Validation 구현

park0503 committed:
refactor: Auth 패키지 Bean Configuration 클래스 분리

park0503 committed:
style: 코드 리포맷팅

park0503 committed:
feat: 예약 대기 중인 스케줄에 예약 대기 요청 시 예외 처리

- 해당 기능 Test 작성

park0503 committed:
feat: 예약 취소 시 다음 순번 대기가 자동으로 예약되도록 기능 수정

- 수정된 기능에 대한 테스트 작성

gayeon1025 commented on src/main/java/auth/JwtTokenProvider.java:
```diff
@@ -1,11 +1,14 @@
 
-import java.util.Date;
-
 @Component
 public class JwtTokenProvider {
```
<Component Scan이 아닌 방법으로 의존성 주입하기> Step1의 요구사항인데 auth 패키지에 @Component, @Service 등의 컴포넌트 스캔 어노테이션들이 남아있는 것 같습니다!

park0503 replied:
component scan에 익숙해진 나머지 그대로 냅두고 말았네요. 지워주겠습니다.
다만 여기서 `@RestController` 어노테이션은 그대로 두는 것이 맞는지 궁금합니다.
컨트롤러의 경우 애초에 Spring 과 밀접하게 연관되어 있기에 Bean으로 수동 등록하는 일이 거의 없는 것으로 알고 있고, `@RestController` 어노테이션을 제거하고 Bean 등록만을 한다면 정상적으로 컨트롤러의 역할을 수행하지 못하는 것 같아서 여쭤봅니다!

gayeon1025 commented on src/main/java/nextstep/reservation/service/ReservationService.java:
```diff
@@ -0,0 +1,88 @@
+        return reservationDao.findAllByThemeIdAndDate(themeId, date);
+    }
+
+    public void deleteById(Member member, Long id) {
+        Reservation reservation = reservationDao.findById(id);
+        if (reservation == null) {
+            throw new RoomReservationException(ErrorCode.RESERVATION_NOT_FOUND);
+        }
+        if (!reservation.sameMember(member)) {
+            throw new RoomReservationException(ErrorCode.RESERVATION_NOT_FOUND);
+        }
+        reservationDao.deleteById(id);
+        ReservationWaiting reservationWaiting = reservationWaitingDao
+                .findFirstByScheduleId(
+                        reservation.getSchedule().getId()
+                );
+        if (Objects.isNull(reservationWaiting)) {
+            return;
+        }
+        reservationWaitingDao.deleteById(reservationWaiting.getId());
+        reservationDao.save(reservationWaiting.convertToReservation());
+    }
```
대기 기능이 생기면서 예약이 취소되면 대기열의 사람을 새로 예약 해주는 로직까지 꼼꼼히 잘 챙겨주셨네요👏
몇가지 함께 봐주셨으면 하는 내용들 커멘트 남겨봅니다.

1. 가능하면 null을 사용하지 않도록 수정해보았으면 해요!
  Optional 키워드를 사용해보면 어떨까 싶습니다

2. transaction 처리
  추가로 더 고민해주셔야할 부분은 transaction인데요~ 
현재 deleteById() 메서드에서는 레파지토리와 관련한 오퍼레이션이 여러번 진행되고 있습니다.

> 1) 예약 삭제
> 2) 예약 대기 삭제
> 3) 신규 예약 생성

위와 같은 순서로 진행되는데, 만약 1), 2) 진행 후에 어떠한 이유로 중단되었다면 첫 번째 대기자가 대기 목록에서 사라지고 예약은 되지 못하는 안타까운 상황이 발생할 것 같아요. 
이런 경우 1), 2), 3)을 하나의 동작처럼 묶어서 동작하도록 해야합니다. 1, 2 번을 정상적으로 끝내고 3번을 수행하닥 실패했다면 수행했던 1,2 번도 되돌리도록(롤백)하도록요!
참고하시면 좋을만한 [블로그](https://kafcamus.tistory.com/30)도 함께 첨부드려 봅니다.

park0503 replied:
1. 이번 과제에서는 스켈레톤 코드도 그렇고 Optional을 따로 사용하지 않아 일단 사용하지 않는 방향으로 구현을 했는데, 역시 null safe 한 코드를 위해서 필요하다고 저 역시도 구현 중간중간 생각을 했었습니다. 전반적인 코드 수정이 필요할 것 같고 제가 지금 따로 3단계를 미리 진행해 놓은 상태라 지금 수정하기 보단 3단계 PR에 적용해서 올리는 것으로 해도 괜찮을까요?

2. 세상에 Transaction을 완전 빼먹고 구현하고 있었네요. 바로 붙여주겠습니다. 다만 이에 대해 갑자기 궁금점이 생겼는데, 저는 보통 서비스 클래스에 `@Transaction` 어노테이션을 달고 조회 쿼리만 이용하는 메소드에 별도로 `@Transaction(readonly = true)`를 붙여주었습니다. 이럴 경우 조회만 하는 경우를 제외하고는 메소드마다 어노테이션을 달지 않아도 됐으나, 쿼리를 1개만 이용하는, 사실상 트랜잭션이 필요 없는 메소드에도 트랜잭션이 적용된다는 점이 갑자기 걸리네요. 실제 필요 여부랑은 별개로 한개의 서비스 메소드는 하나의 트랜잭션과 생명주기가 일치해야한다고 생각하기에 명시적으로 표현해준다는 점도 있어서 원래는 클래스에 어노테이션을 달았던 건데, 이에 관해 카야의 생각이 궁금합니다!

gayeon1025 replied:
> 1. 전반적인 코드 수정이 필요할 것 같고 제가 지금 따로 3단계를 미리 진행해 놓은 상태라 지금 수정하기 보단 3단계 PR에 적용해서 올리는 것으로 해도 괜찮을까요?

Q1. 넵 좋습니다!

> 2. 실제 필요 여부랑은 별개로 한개의 서비스 메소드는 하나의 트랜잭션과 생명주기가 일치해야한다고 생각하기에 명시적으로 표현해준다는 점도 있어서 원래는 클래스에 어노테이션을 달았던 건데, 이에 관해 카야의 생각이 궁금합니다!

Q2. 넵 저도 죠스의 생각에 동의합니다! DB 접근은 모두 트랜잭션 내에서 이루어져야 acid를 보장한다고할 수 있을 것 같아요. 싱글 쿼리라도 수정이 일어나는 쿼리는 롤백을 위해서 transactional 처리가 필요하므로 클래스에 붙여주셔도 될 것 같습니다.

gayeon1025 commented on src/main/java/nextstep/reservation/service/ReservationWaitingService.java:
```diff
@@ -0,0 +1,66 @@
+        this.reservationService = reservationService;
+    }
+
+    public Long create(Member member, ReservationRequest reservationRequest) {
+        Schedule schedule = scheduleDao.findById(reservationRequest.getScheduleId());
+        if (Objects.isNull(schedule)) {
+            throw new RoomReservationException(ErrorCode.SCHEDULE_NOT_FOUND);
+        }
+        List<Reservation> reservationList = reservationDao.findByScheduleId(schedule.getId());
+        if (reservationList.size() == 0) {
+            reservationService.create(member, reservationRequest);
+        }
+        List<ReservationWaiting> reservationWaitingList = reservationWaitingDao.findByScheduleId(schedule.getId());
+        boolean isDuplicated = reservationWaitingList.stream().anyMatch(reservationWaiting -> reservationWaiting.isMine(member));
+        if (isDuplicated) {
+            throw new RoomReservationException(ErrorCode.DUPLICATE_RESERVATION_WAITING);
+        }
+        long waitNum = reservationWaitingList.isEmpty() ? 1
+                : reservationWaitingList.get(reservationWaitingList.size() - 1).getWaitingNum() + 1;
+        return reservationWaitingDao.save(new ReservationWaiting(schedule, member, waitNum));
+    }
```
1. 일관된 예외처리 및 null handling 
    ```
    if (Objects.isNull(schedule)) {
        throw new RoomReservationException(ErrorCode.SCHEDULE_NOT_FOUND);
    }
    ```
    전체 코드에서 null을 확인하는 부분이 조금씩 다른 것 같아요! `if (reservationWaiting == null)`  도 있고 위와 같이 Objects.isNull(...)을 활용하기도 하네요. 통일되면 좋을 것 같아요! null 키워드 대신 앞에서 언급드린 대로 Optional을 활용하면 어떨까요?

2. 로직 점검
    ```
     if (reservationList.size() == 0) {
          reservationService.create(member, reservationRequest);
      }
    ```
    요구 사항 중 [예약이 없는 스케줄에 대해서 예약 대기 신청을 할 경우 예약이 된다.] 라는 부분을 구현해주신 것 같은데요.
    만약 예약이 없어서 예약을 진행했다면, 그 이후에 예약 대기를 생성하는 부분은 진행하지 않아도 될 것 같아요. 

3. 순번 관리 ⭐️ (여기서 가장 고민해봤으면 하는 부분입니다!)
 아마 들어보셨겠지만 해당 메서드에서는 **스레드 간의 경쟁**이 발생될 것으로 우려됩니다.
스레드 A가 먼저 로직을 수행하던 중 완료되기 전에 스레드 B도 로직을 수행하게 되면, 
 `List<ReservationWaiting> reservationWaitingList = reservationWaitingDao.findByScheduleId(schedule.getId());` 에서 아직 A가 저장을 하기 전이기 때문에 waitNum을 우리가 기대했던 값과는 다른 값을 갖게 될 것 같아요.
저는 해결방법이 2가지정도 떠오르는데 혹시 죠스도 생각나는 방법이 있을까요? 까다로운 부분이라 혹시나 너무 어렵고 생각이 나지 않는다면 편하게 말씀주세요!!

park0503 replied:
1. 위에 말씀하신대로 Optional을 이용하면 해결될 것 같습니다. Optional 적용 전에는 일단 표현 방법을 Objects.isNull, Objects.nonNull 메소드를 이용하는 것으로 통일해 놓겠습니다.
2. return을 빼먹었군요. 당장 수정하겠습니다.
3. 멀티 쓰레딩 환경을 생각하지 못했군요! 일단 당장은 쿼리를 나릴 때 for update를 붙여서 데이터베이스 레벨에서 배타락을 거는 방법이 생각났는데, 이론만 배웠고 실제로 사용해본 적은 없어서 맞는지 확신이 잘 안서네요. synchronized 키워드를 사용하는것도 가능한가 해서 찾아보니 Transaction 어노테이션과는 함께 사용할 수 없어 보여 현재는 db의 lock을 이용한 방법만이 생각납니다.

park0503 replied:
~~3번의 또다른 방법으로, 해당 메소드에서만 트랜잭션 격리 수준을 serializable로 설정하는 것도 될 수 있을까요? 성능저하가 심한 격리수준이라 실제로 거의 안쓰인다고 들은 터라 맞는 방법인지는 잘 모르겠네요 ㅎㅎ,,,~~
serializable 격리수준도 결국 shared lock만 걸어주는 거라 해결책이 되기는 힘들 것 같네요 ㅜㅜ

gayeon1025 replied:
> 일단 당장은 쿼리를 나릴 때 for update를 붙여서 데이터베이스 레벨에서 배타락을 거는 방법이 생각났는데, 이론만 배웠고 실제로 사용해본 적은 없어서 맞는지 확신이 잘 안서네요. synchronized 키워드를 사용하는것도 가능한가 해서 찾아보니 Transaction 어노테이션과는 함께 사용할 수 없어 보여 현재는 db의 lock을 이용한 방법만이 생각납니다.

[A 스레드가 로직을 수행하는 중(커밋은 이루어지지 않았음)에 B 스레드가 read를 하는 상황]
1. 낙관적 락을 활용하기 (📌[참고](https://isntyet.github.io/jpa/JPA-%EB%82%99%EA%B4%80%EC%A0%81-%EC%9E%A0%EA%B8%88(Optimisstic-Lock)/))
  JPA에서 제공하는 @Version 필드를 활용합니다. A스레드가 커밋 후에 B 스레드가 커밋하려고 한다면 ObjectOptimisticLockingFailureException 이 발생합니다. 이를 catch 하여 처리.

2. 비관적 락을 활용하기 (📌 [참고](https://isntyet.github.io/jpa/JPA-%EB%B9%84%EA%B4%80%EC%A0%81-%EC%9E%A0%EA%B8%88(Pessimistic-Lock)/))
  비관적락을 설정한 쿼리는 다른 트랜잭션에서 읽지도 쓰지도 못하게 한다.(이게 exclusive lock에 해당하겠지요!) 안전해보이지만 성능저하라는 trade-off가 있을 수 있고, 데드락이 발생할 수 있으니 유의해야 합니다.

3. AtomicInteger 활용하기
 이게 가장 간단한 방법인데요~ 클래스 멤버로 AtomicInteger로 number를 선언 후에 사용하면 될 것 같아요. 다만 이 방법은 싱글 어플리케이션에서만 활용할 수 있는 방법입니다. 스케일 아웃을 할 경우 각 어플리케이션마다 number를 관리하기 때문에 충돌이 발생할 수 있는 방법입니다! 지금은 싱글 어플리케이션이니 AtomicInteger을 사용해보셔도 좋을 것 같습니다!
(+ 스케일 아웃 환경에서는 레디스 등을 활용합니다! 궁금하시다면 '레디스를 활용한 분산락' 등의 키워드로 검색해보시면 좋은 자료들이 많을 것 같아요)

저도 항상 어려워하는 부분이라;; 제가 아는 바대로 열심히 적어봤습니다! 혹시 관련해서 서칭하시다가 제가 언급한 부분 중에 틀린 부분이 있다면 언제든 말씀주세요! 위 방법중 죠스가 해보고 싶은 방식으로 채택해주시면 될 것 같습니다~


📌 낙관적락 vs 비관적락 ([link](https://jaehoney.tistory.com/159))

gayeon1025 reviewed (changes requested):
안녕하세요 죠스! 
요구사항 구현하시라 고생 많으셨습니다~ 평소처럼 잘 하셨지만 추가로 고민해주셨으면 하는 부분들 커멘트 남겨봤습니다. 

1단계 리팩토링하면서 auth 패키지로 분리해주셨는데, auth 패키지 내부에서도 interceptor, exception, service 등과 같이 sub-package로 구분해주시면 좋을 것 같습니다!

park0503 committed:
refactor: 피드백 반영

- auth 패키지 하위 세분화
- auth 패키지 내 컴포넌트 스캔용 어노테이션 제거
- null 검증 임시 획일화
- Transaction 적용
- 예약 대기 생성 로직 오류 수정

park0503 commented:
> 안녕하세요 죠스! 요구사항 구현하시라 고생 많으셨습니다~ 평소처럼 잘 하셨지만 추가로 고민해주셨으면 하는 부분들 커멘트 남겨봤습니다.
> 
> 1단계 리팩토링하면서 auth 패키지로 분리해주셨는데, auth 패키지 내부에서도 interceptor, exception, service 등과 같이 sub-package로 구분해주시면 좋을 것 같습니다!

일단 피드백 일부 반영해 커밋했습니다!
트랜잭션의 경우 앞서 말한 고민이 있기에 이번에는 필요한 메소드에 어노테이션을 달아주는 것으로 진행했습니다.
순번관리의 레이스 컨디션 문제에 대해서는 조금 더 리서치를 진행해 볼 계획입니다!

park0503 committed:
refactor: critical section에 배타락 도입

- ReservationWaitingDao -> findByScheduleId
  - ReservationWaiting 생성 시 대기번호 산정 로직에 필요
  - Schedule 삭제 시 보유 예약 확인 로직에 필요
- ScheduleDao -> findByThemeId
  - Theme 삭제 시 보유 스케줄 확인 로직에 필요

gayeon1025 reviewed (changes requested):
죠스! shared lock, exclusive lock 모두 공부하셨다니 👍
저도 'for update'라는 키워드는 덕분에 처음 알았어요! 

[critical section에 배타락 도입] 커밋 내용에 대한 리뷰를 추가로 드립니다.
1. 쿼리에 'for update' 키워드가 있을 경우 db에 의존적인 코드가 될 수 있습니다.
  'for update' 코드가 제가 커멘트 드린 '비관적락'과 동일한 것 같으니 참고 부탁드릴게요~
2. 'for update'의 기본 옵션을 사용하고 있어 무한정 대기에 빠질 수 있습니다.
 저도 덕분에 찾다보니 'for update'에 [옵션](https://grandma-coding.tistory.com/entry/MySQL-SELECT-FOR-UPDATE-%EB%9E%80)이 있더라구요! 현재는 아무 옵션도 없어서 무한정 대기 상태에 있으니 적절히 timeout을 하여 exception을 핸들링하는 방법이 더 좋아보입니다!

park0503 committed:
refactor: 트랜잭션 어노테이션 클래스에 적용

park0503 committed:
refactor: 대기번호 산정 동기화 로직 변경

park0503 committed:
feat: Lock 적용

- 예약 대기 번호 산정
- 스케줄 삭제 시 잔여 예약 검사
- 테마 삭제 시 잔여 스케줄 검사

park0503 commented:
말씀해주신 방법들 잘 읽었습니다 카야! 확실히 배타락 방식은 DB 쿼리를 직접 변경하는 것이기 때문에 말씀대로 DB에 의존적인 애플리케이션이 될 수 있겠네요. 따라서 일단은 말씀해주신 AtomicInteger를 이용해 동기화 처리를 해주었습니다. Lock처럼 이용하기 위해 AtomicInteger가 0이면 Unlock, 1이면 Lock 상태라고 생각하고 구현했습니다.

다만 다소 불안한 부분이  unLock 부분인데요, 중간에 예외가 터지는 경우의 unLock에 문제가 발생할 수도 있다는 생각이 듭니다. 지금은 제가 throw를 해주는 부분 전에는 unLock 을 따로 해 준 상태지만 예상하지 못한 곳에서 예외가 터지면 unLock이 되지 않기에 여기서 데드락이 발생할 수도 있지 않나는 생각이 듭니다. 어렵네요... ㅎㅎ

park0503 committed:
refactor: Lock 클래스 도입

park0503 commented:
가독성을 위해 Lock 클래스를 구현하고 도입했습니다.

gayeon1025 commented:
@park0503 
죠스 엄청 고민의 흔적이 보이네요!! 👏👏👏
저는 그냥 대기순번을 AtomicInteger로 작성해서 완전 간단하게 생각했는뎈ㅋㅋㅋ 뮤텍스를 들고오셨네요!! (저도 덕분에 다시 공부했습니다..ㅎ)
`HashMap<Long, AtomicInteger> waitingNum`(<scheduleId, 순번> 의 쌍 )  을 그냥 선언해두고 저기서 순번을 꺼내오는 걸로 생각했었어요

저도 덕분에 동시성 문제를 해결하는 방법이나 trasaction, db lock 등을 다시 한 번 생각할 수 있는 기회였습니다.
현재 구현의 한계까지 생각해봐주셔서 제가 기대했던 것 이상으로 잘 해주신 것 같습니다.

이상 머지하겠습니다!
